tests/MongoDB.Driver.Tests/Specifications/Runner/MongoClientJsonDrivenTestRunnerBase.cs

```diff
@@ -0,0 +1,392 @@
+            return DriverTestConfiguration.CreateDisposableClient(settings =>
+            {
+                ConfigureClientSettings(settings, test);
+                if (eventCapturer != null)
+                {
+                    settings.ClusterConfigurator = c => c.Subscribe(eventCapturer);
+                }
+            });
+        }
+
+        protected void DropCollection()
+        {
+            var client = DriverTestConfiguration.Client;
+            var database = client.GetDatabase(DatabaseName).WithWriteConcern(WriteConcern.WMajority);
+            database.DropCollection(CollectionName);
+        }
+
+        protected void InsertData(BsonDocument shared)
+        {
+            if (shared.Contains(DataKey))
+            {
+                var documents = shared[DataKey].AsBsonArray.Cast<BsonDocument>().ToList();
+                if (documents.Count > 0)
+                {
+                    var client = DriverTestConfiguration.Client;
+                    var database = client.GetDatabase(DatabaseName);
+                    var collection = database.GetCollection<BsonDocument>(CollectionName).WithWriteConcern(WriteConcern.WMajority);
+                    collection.InsertMany(documents);
+                }
+            }
+        }
+
+        protected void RunTests(JsonDrivenTestCase testCase)
+        {
+            CheckServerRequirements(testCase.Shared);
+            Run(testCase.Shared, testCase.Test);
+        }
+
+        protected void VerifyCollectionData(IEnumerable<BsonDocument> expectedDocuments)
+        {
+            var database = DriverTestConfiguration.Client.GetDatabase(DatabaseName);
+            var collection = database.GetCollection<BsonDocument>(CollectionName);
+            var actualDocuments = collection.Find("{}").ToList();
+            actualDocuments.Should().BeEquivalentTo(expectedDocuments);
+        }
+
+
+        // private methods
+        private void AssertEvent(object actualEvent, BsonDocument expectedEvent)
+        {
+            if (expectedEvent.ElementCount != 1)
+            {
+                throw new FormatException("Expected event must be a document with a single element with a name the specifies the type of the event.");
+            }
+
+            var eventType = expectedEvent.GetElement(0).Name;
+            var eventAsserter = EventAsserterFactory.CreateAsserter(eventType);
+            eventAsserter.AssertAspects(actualEvent, expectedEvent[0].AsBsonDocument);
+        }
+
+        private ReadPreference ReadPreferenceFromBsonValue(BsonValue value)
+        {
+            if (value.BsonType == BsonType.String)
+            {
+                var mode = (ReadPreferenceMode)Enum.Parse(typeof(ReadPreferenceMode), value.AsString, ignoreCase: true);
+                return new ReadPreference(mode);
+            }
+
+            return ReadPreference.FromBsonDocument(value.AsBsonDocument);
+        }
+
+        private void Run(BsonDocument shared, BsonDocument test)
```

Comment:
Let's name this method

    SetupAndRunTest

Reply:
Done